llms/mlx_lm/tuner/dora.py

```diff
@@ -75,16 +83,57 @@ def __init__(
             shape=(input_dims, r),
         )
         self.lora_b = mx.zeros(shape=(r, output_dims))
-        self.m = mx.linalg.norm(self.linear.weight, axis=1)
+
+    def set_linear(self, linear):
+        """
+        Set the self.linear layer and recompute self.m with respect to quantization
+        """
+        self.linear = linear
+        # self.m needs to be recomputed when self.linear changes
+        # which doesn't happen in from_linear when dora_lin.linear = linear
+        self.m = mx.linalg.norm(self._dequantized_weight(), axis=1).astype(mx.float32)
+
+    def _dequantized_weight(self):
+        """
+        Return the weight of linear layer and dequantize it if is quantized
+        """
+        weight = self.linear.weight
+        if self._is_quantized():
+            weight = mx.dequantize(
+                weight,
+                self.linear.scales,
+                self.linear.biases,
+                self.linear.group_size,
+                self.linear.bits,
+            )
+        return weight
+
+    def _is_quantized(self):
+        return isinstance(self.linear, nn.QuantizedLinear)
 
     def __call__(self, x):
         # Regular LoRA (without a bias)
-        y = x @ self.linear.weight.T
+        if self._is_quantized():
+            # Use quantized_matmul instead of dequantizing for efficiency
+            y = mx.quantized_matmul(
+                x,
+                self.linear.weight,
+                scales=self.linear.scales,
+                biases=self.linear.biases,
+                transpose=True,
+                group_size=self.linear.group_size,
+                bits=self.linear.bits,
+            )
+        else:
+            y = x @ self.linear.weight.T
+
         z = (self.dropout(x) @ self.lora_a) @ self.lora_b
         out = y + (self.scale * z).astype(x.dtype)
 
         # Compute the norm of the adapted weights
-        adapted = self.linear.weight + (self.scale * self.lora_b.T) @ self.lora_a.T
+        adapted = (
+            self._dequantized_weight() + (self.scale * self.lora_b.T) @ self.lora_a.T
+        )
```

Comment:
We need to be a bit careful with the types here, I think there is a existing bug. We want the output of this function to have the same type as the input `x`. So we should probably cast the scale `(self.m / denom)` to the type of the output. E.g. change line 140 to:

```
out = (self.m / denom).astype(out.dtype) * out
```